sdk-dotnet/src/LittleHorseSDK/LittleHorse.Worker/Internal/LHServerConnectionManager.cs

```diff
@@ -0,0 +1,260 @@
+        private void RebalanceWork()
+        {
+            while (_running)
+            {
+                DoHeartBeat();
+                try
+                {
+                    Thread.Sleep(BALANCER_SLEEP_TIME);
+                }
+                catch { }
+            }
+        }
+
+        private void DoHeartBeat()
+        {
+            try
+            {
+                var request = new RegisterTaskWorkerRequest()
+                {
+                    TaskDefName = _taskDef.Name,
+                    ClientId = _config.ClientId,
+                    ListenerName = _config.ConnectListener
+                };
+
+                var response = _bootstrapClient.RegisterTaskWorker(request);
+                
+                HandleRegisterTaskWorkResponse(response);
+
+            } 
+            catch (Exception ex)
+            {
+                _logger?.LogError(ex, $"Failed contacting bootstrap host {_config.APIBootstrapHost}:{_config.APIBootstrapPort}");
+            }
+        }
+
+        private void HandleRegisterTaskWorkResponse(RegisterTaskWorkerResponse response)
+        {
+
+            response.YourHosts.ToList().ForEach(async host =>
+            {
+                if(!IsAlreadyRunning(host))
+                {
+                    try
+                    {
+                        var newConnection = new LHServerConnection<T>(this, host);
+                        await newConnection.Connect();
+                        _runningConnections.Add(newConnection);
+                        _logger?.LogInformation($"Adding connection to: {host.Host} : {host.Port} for {_taskDef.Name}");
+                    }
+                    catch (IOException ex)
+                    {
+                        _logger?.LogError(ex, $"Exception on HandleRegisterTaskWorkResponse.");
+                    }
+                }
+                
+            });
+
+            var lastIndexOfRunningConnection = _runningConnections.Count() - 1;
+
+            for (int i = lastIndexOfRunningConnection; i >= 0; i--) 
+            { 
+                var runningThread = _runningConnections[i];
+
+                if(!ShouldBeRunning(runningThread, response.YourHosts))
+                {
+                    _logger?.LogInformation($"Stopping worker thread for host {runningThread.HostInfo.Host} : {runningThread.HostInfo.Port}");
+
+                    runningThread.Dispose();
+                    _runningConnections.RemoveAt(i);
+                }
+            }
+        }
+
+        private bool ShouldBeRunning(LHServerConnection<T> runningThread, RepeatedField<LHHostInfo> hosts)
+        {
+            return hosts.ToList().Any(host => runningThread.IsSame(host));
+        }
+
+        private bool IsAlreadyRunning(LHHostInfo host)
+        {
+            return _runningConnections.Any(conn => conn.IsSame(host));
+        }
+
+        public async void SubmitTaskForExecution(ScheduledTask scheduledTask, LHPublicApi.LHPublicApiClient client)
```

Comment:
Since this is async, it seems like the worker threads will poll tons and tons of tasks and then queue them up in a bunch of calls that are waiting on the `_semaphore`. This is a hazard because the task will go  from `SCHEDULED` to `RUNNING`, and then it might time out before it starts being executed.